lib/ddtrace/contrib/rack/rum_injection.rb

```diff
@@ -0,0 +1,273 @@
+          injectable = should_inject?(headers, env, status)
+
+          if injectable
+            current_trace_id = get_current_trace_id
+
+            # do not inject if no trace
+            return result unless current_trace_id
+
+            # we need to insert the trace_id and expiry meta tags
+            updated_html = generate_updated_html(response, headers, current_trace_id)
+
+            # update content length and return new response if we don't fail on rum injection
+            if updated_html
+              update_content_length(headers, updated_html)
+              # ensure idempotency on injection in case middleware is inserted or called twice
+              env[RUM_INJECTION_FLAG] = true
+
+              updated_response = ::Rack::Response.new(updated_html, status, headers)
+              Datadog.logger.debug('Rum injection successful')
+              return updated_response.finish
+            else
+              Datadog.logger.debug('Rum injection unsuccessful')
+              return result
+            end
+          end
+
+          # catchall if an earlier conditional is not met
+          result
+        rescue Exception => e # rubocop:disable Lint/RescueException
+          Datadog.logger.warn("error checking rum injectability #{e.class}: #{e.message} #{e.backtrace.join("\n")}")
+          raise e
+        end
+
+        private
+
+        def configuration
+          Datadog.configuration[:rack]
+        end
+
+        def should_inject?(headers, env, status)
+          status == 200 &&
+            !env[RUM_INJECTION_FLAG] &&
+            no_cache?(headers, env) &&
+            !compressed?(headers) &&
+            !attachment?(headers) &&
+            !streaming?(headers, env) &&
+            injectable_html?(headers)
+        # catch everything and swallow it here for defensiveness
+        rescue Exception => e # rubocop:disable Lint/RescueException
+          Datadog.logger.warn("Error during rum injection  #{e.class}: #{e.message} #{e.backtrace.join("\n")}")
+          return nil
+        end
+
+        def compressed?(headers)
+          headers && headers.key?(CONTENT_ENCODING_HEADER) &&
+            (headers[CONTENT_ENCODING_HEADER].include?(COMPRESS) ||
+              headers[CONTENT_ENCODING_HEADER].include?(GZIP) ||
+              headers[CONTENT_ENCODING_HEADER].include?(DEFLATE))
+        end
+
+        def injectable_html?(headers)
+          (headers && headers.key?(CONTENT_TYPE_HEADER) && !headers[CONTENT_TYPE_HEADER].nil?) &&
+            (headers[CONTENT_TYPE_HEADER].include?(HTML_CONTENT) ||
+            headers[CONTENT_TYPE_HEADER].include?(XHTML_CONTENT))
+        end
+
+        def attachment?(headers)
+          (headers && headers.key?(CONTENT_DISPOSITION_HEADER) && !headers[CONTENT_DISPOSITION_HEADER].nil?) &&
+            headers[CONTENT_DISPOSITION_HEADER].include?(ATTACHMENT)
+        end
+
+        def streaming?(headers, env)
```

Comment:
If my suggestion for wrapping the response body object, and prepending only a single line to the body is accepted, we can actually support streaming, thus this method would become unnecessary.

Reply:
for the POC it was decided not to support streaming.

Additionally rails has some pretty clear documentation around not modifying response bodies in rack middleware when streaming and that they don't intended to support middlewares that do that.